torch/distributed/_spmd/api.py

```diff
@@ -143,7 +143,7 @@ def _rematerialize_optimizer(
     assert opt is not None
 
     # update opt.state with proxy tensors
-    orig_states: Dict[str, Any] = copy(opt.state)
+    orig_states = copy(opt.state)
```

Comment:
Because I hinted `state` as `DefaultDict` we get:

```
torch/distributed/_spmd/api.py:160:21: error: Incompatible types in assignment (expression has type "Dict[str, Any]", variable has type "defaultdict[str, Dict[Any, Any]]")  [assignment]
```

But also yes, typically I only really see generic types (like `list`, `dict`, etc.) annotated when they are initialized as empty or the type is more general than what you are initializing it to. Here we can just infer the type from the copy.

Reply:
ah. this hints at bigger problems I guess--can we constrain opt.state to be a defaultdict without breaking old code like this?

Reply:
Sorry, I had reversed the order in my previous comment. My concern is that if this code (in distributed) made assumptions and added the type for the result of `copy(opt.state)` to be `Dict[str, Any]` and had no problems before but would run into problems now, then there may be user code that do the same thing today and will be broken by this change.

Reply:
I understand your concern about breaking things in user code, but I think what we see here is about as much as we can expect user code to "break". This change was purely typing related. Nothing has changed about how this code would run. It was a `defaultdict` before and it still is now, and `defaultdict` is still just a subclass of `dict`.

Reply:
Ah, I see state was always a defaultdict, the contention here is not Dict vs DefaultDict, but the value of the dict being Any vs Dict[Any, Any], which doesn't work out with typechecking cuz of contravariance, I suppose?

Reply:
Well... pyright doesn't complain that `DefaultDict[str, Dict[Any, Any]]` is being casted down to `Dict[str, Any]`... this seems like a quirk of `mypy` IMO. it is really weirdly strict about this kind of assignments.